File: convert-png-to-webp.py
import sys
from pathlib import Path

try:
    from PIL import Image
except ImportError:
    print("Error: Pillow module not found.")
    print("Please install it by running: python -m pip install --user Pillow")
    sys.exit(1)

# 사용법 안내
if len(sys.argv) < 2:
    print("사용법:")
    print("  python convert-png-to-webp.py <png_폴더_경로>")
    print("예시:")
    print("  python convert-png-to-webp.py assets/games/weapon-levelup/images")
    sys.exit(1)

# 인자로 받은 폴더(=png가 있는 images 폴더)
source_dir = Path(sys.argv[1]).resolve()

if not source_dir.exists() or not source_dir.is_dir():
    print("폴더를 찾을 수 없습니다:", source_dir)
    sys.exit(1)

# 출력 폴더: images/webp
webp_dir = source_dir / "webp"
webp_dir.mkdir(exist_ok=True)
print(f"Source directory: {source_dir}")
print(f"Output directory: {webp_dir}")

# PNG 파일 찾기(대소문자 모두)
png_files = list(source_dir.glob("*.png")) + list(source_dir.glob("*.PNG"))

if not png_files:
    print("No PNG files found in the source directory.")
    sys.exit(0)

print(f"Found {len(png_files)} PNG file(s) to convert.\n")

success_count = 0
error_count = 0

for png_file in png_files:
    try:
        img = Image.open(png_file)

        # 투명도 보존용
        if img.mode != "RGBA":
            img = img.convert("RGBA")

        webp_file = webp_dir / f"{png_file.stem}.webp"

        # 이미 존재하면 덮어쓰기(원치 않으면 여기서 skip 처리 가능)
        img.save(webp_file, "WEBP", quality=80, method=6)

        original_size = png_file.stat().st_size
        webp_size = webp_file.stat().st_size
        size_reduction = (1 - webp_size / original_size) * 100 if original_size else 0

        print(f"✓ {png_file.name} → webp/{webp_file.name} ({size_reduction:.1f}% smaller)")
        success_count += 1

    except Exception as e:
        print(f"✗ Error converting {png_file.name}: {e}")
        error_count += 1

print("\nConversion complete!")
print(f"Success: {success_count} file(s)")
if error_count > 0:
    print(f"Errors: {error_count} file(s)")

File: convert-png-to-webp.js
/**
 * PNG to WebP Converter
 * Converts all PNG files in assets/games/weapon-levelup/images to WebP format
 * - Quality: 80
 * - Transparency: Preserved
 * - Original PNG files: Kept
 * - WebP files: Saved to /webp subfolder
 */

const fs = require('fs');
const path = require('path');

// Check if sharp is available
let sharp;
try {
  sharp = require('sharp');
} catch (err) {
  console.error('Error: sharp module not found.');
  console.error('Please install it by running: npm install sharp');
  process.exit(1);
}

const sourceDir = path.join(__dirname, 'assets', 'games', 'weapon-levelup', 'images');
const webpDir = path.join(sourceDir, 'webp');

// Create webp directory if it doesn't exist
if (!fs.existsSync(webpDir)) {
  fs.mkdirSync(webpDir, { recursive: true });
  console.log(`Created directory: ${webpDir}`);
}

// Get all PNG files
const files = fs.readdirSync(sourceDir).filter(file => 
  file.toLowerCase().endsWith('.png')
);

if (files.length === 0) {
  console.log('No PNG files found in the source directory.');
  process.exit(0);
}

console.log(`Found ${files.length} PNG file(s) to convert.\n`);

let successCount = 0;
let errorCount = 0;

// Convert each PNG file
async function convertFiles() {
  for (const file of files) {
    const inputPath = path.join(sourceDir, file);
    const outputFileName = file.replace(/\.png$/i, '.webp');
    const outputPath = path.join(webpDir, outputFileName);
    
    try {
      await sharp(inputPath)
        .webp({ 
          quality: 80,
          effort: 6 // 0-6, higher = better compression but slower
        })
        .toFile(outputPath);
      
      const inputStats = fs.statSync(inputPath);
      const outputStats = fs.statSync(outputPath);
      const sizeReduction = ((1 - outputStats.size / inputStats.size) * 100).toFixed(1);
      
      console.log(`✓ ${file} → ${outputFileName} (${sizeReduction}% smaller)`);
      successCount++;
    } catch (error) {
      console.error(`✗ Error converting ${file}:`, error.message);
      errorCount++;
    }
  }
  
  console.log(`\nConversion complete!`);
  console.log(`Success: ${successCount} file(s)`);
  if (errorCount > 0) {
    console.log(`Errors: ${errorCount} file(s)`);
  }
}

convertFiles().catch(err => {
  console.error('Fatal error:', err);
  process.exit(1);
});

